program/src/processor/slash.rs

```diff
@@ -11,14 +24,187 @@ use crate::instruction::accounts::{Context, SlashAccounts};
 /// 1. `[w]` Stake account
 /// 2. `[s]` Slash authority
 /// 3. `[w]` Vault token account
-/// 4. `[]` Vault authority, PDA with seeds `['token-owner', stake_config]`
-/// 5. `[]` SPL Token program
+/// 4. `[]` Stake Token Mint
+/// 5. `[]` Vault authority, PDA with seeds `['token-owner', stake_config]`
+/// 6. `[]` SPL Token program
 ///
 /// Instruction data: amount of tokens to slash
 pub fn process_slash(
-    _program_id: &Pubkey,
-    _ctx: Context<SlashAccounts>,
-    _amount: u64,
+    program_id: &Pubkey,
+    ctx: Context<SlashAccounts>,
+    amount: u64,
 ) -> ProgramResult {
-    Ok(())
+    // Account validation.
+
+    // config
+    // - owner must be the stake program
+    // - must be initialized
+
+    require!(
+        ctx.accounts.config.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "config"
+    );
+
+    let mut config_data = ctx.accounts.config.try_borrow_mut_data()?;
+    let config = bytemuck::try_from_bytes_mut::<Config>(&mut config_data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        config.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "config",
+    );
+
+    // stake
+    // - owner must be the stake program
+    // - must be initialized
+    // - derivation must match (validates the config account)
+
+    require!(
+        ctx.accounts.stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "stake"
+    );
+
+    let mut stake_data = ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(&mut stake_data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    let (derivation, _) =
+        find_stake_pda(&stake.validator_vote, ctx.accounts.config.key, program_id);
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // slash authority
+    // - must be a signer
+    // - must match the slash authority on the config account
+    //
+    // When there is no slash authority set, the stake account cannot be slashed and
+    // an error is returned.
+
+    let slash_authority = config.slash_authority.into();
+
+    if let Some(slash_authority) = slash_authority {
```

Comment:
Up to you, but I think you can one-line this.

```suggestion
    if let Some(slash_authority) = Option::<Pubkey>::from(config.slash_authority) {
```

Reply:
Nice!